mlir/lib/Quantum/Transforms/AdjointPatterns.cpp

```diff
@@ -264,6 +268,96 @@ class AdjointGenerator {
+            getQuantumReg(funcOp.front().getTerminator()->getOperands());
+        if (!returnedQureg.has_value()) {
+            // This operation is purely classical
+            return;
+        }
+
+        // Save the insertion point to come back to it after creating the adjoint of the function
+        auto insertionSaved = builder.saveInsertionPoint();
+        MLIRContext *ctx = builder.getContext();
+
+        // Create the adjoint of the original function at the module level
+        ModuleOp moduleOp = funcOp->getParentOfType<ModuleOp>();
+        builder.setInsertionPointToStart(moduleOp.getBody());
+
+        Block *originalBlock = &funcOp.front();
+        Operation *originalTerminator = originalBlock->getTerminator();
+        ValueRange originalArguments = originalBlock->getArguments();
+
+        // Get the arguments and outputs types from the original function (same signature)
+        FunctionType adjointFnType =
+            FunctionType::get(ctx, /*inputs=*/
+                              originalArguments.getTypes(),
+                              /*outputs=*/originalTerminator->getOperandTypes());
+        std::string adjointName = funcOp.getName().str() + ".adjoint";
+        Location loc = funcOp.getLoc();
+        func::FuncOp adjointFnOp = builder.create<func::FuncOp>(loc, adjointName, adjointFnType);
+        adjointFnOp.setPrivate();
+
+        // Create the block of the adjoint function
+        Block *adjointFnBlock = adjointFnOp.addEntryBlock();
+        builder.setInsertionPointToStart(adjointFnBlock);
+        Type qregType = quantum::QuregType::get(builder.getContext());
+        auto argumentsSize = adjointFnOp.getArguments().size();
+
+        // The last argument is the quantum register
+        Value lastArg = adjointFnOp.getArgument(argumentsSize - 1);
+        assert(isa<QuregType>(lastArg.getType()) &&
+               "The last argument of the function must be the quantum register.");
+        quantum::AdjointOp adjointOp = builder.create<quantum::AdjointOp>(loc, qregType, lastArg);
+
+        Region *adjointRegion = &adjointOp.getRegion();
+        Region &originalRegion = funcOp.getRegion();
+
+        // Map the arguments with the original function
+        IRMapping map;
+        for (size_t i = 0; i < funcOp.getArguments().size() - 1; i++) {
+            Value arg = adjointFnOp.front().getArgument(i);
+            Value argBlock = originalRegion.front().getArgument(i);
+            map.map(argBlock, arg);
+        }
+
+        // Copy the original region in the adjoint region
+        originalRegion.cloneInto(adjointRegion, map);
+
+        // Replace the return of the quantum register with a quantum yield
+        auto terminator = adjointRegion->front().getTerminator();
+        ValueRange res = terminator->getOperands();
+        TypeRange resTypes = terminator->getResultTypes();
+        builder.setInsertionPointAfter(terminator);
+        builder.create<quantum::YieldOp>(loc, resTypes, res);
```

Comment:
Just got excited because it looked like the pass handles functions that return both a register and classical results. But looking at the adjoint op, this isn't possible (the only return value is a quantum.reg).

I don't know if there is a use case for it or not, but I guess if we allowed the AdjointOp to return arbitrary values, it would allow values that are defined inside the adjoint region to be accessible outside of it. Not really needed under normal conditions, but with the addition of nested function calls I could see a function returning classical values in addition to a quantum register.

Have you thought about whether we can handle this case? I know it's not needed for the ZNE transform, but even just being aware whether it's easy or hard to support would be useful. 
At a minimum though, we need to add an error checking for presence of additional results.

(The considerations above don't apply to arguments, as values can be freely accessed inside the adjoint region from the outer scope.)

Reply:
Yes I think we can handle the case, but it makes it more complicated. The forward pass must be changed like we do for ForOp and WhileOp. You need to separate the classical part and quantum part, the later is kept for the backward pass. Happy to revisit if it becomes useful in the future.

Reply:
I think it was done there https://github.com/PennyLaneAI/catalyst/pull/353/commits/0fbfa96334a6629d54df2b587e6f64c9f48c3f04

Reply:
Oh thanks, just noticed it too 👍